### ISSUE TITLE
Register NLog configuration items without assembly reflection

### ISSUE THREAD

**Issue by snakefoot** [NLog/NLog#5190]
Instead of using reflection and class attributes, then it should be possible to register NLog Configuration Item Types directly:
- NLog Target
- NLog Layout
- NLog LayoutRenderer (Along with ambient properties)
- NLog Conditions
- NLog Filters
- NLog TimeSource

When having introduced the new system, then all default NLog Configuration Items (in `NLog.dll` + `NLog.Extensions.Logging` + `NLog.Web` + `NLog.Web.AspNetCore`) should use the new system, and only use the old reflection-based-approach for backwards-compatibility (Ex. loading assemblies with NLog-extensions).

I guess that means loosing the multi-type-alias-logic, since the obvious type-alias would be the class-name. And to have multiple-type-alias then one would just have inheritance-class-name for each alias.

I guess it also means obsoleting the old dependency injection override logic, where re-assigning [ConfigurationItemFactory.CreateInstance](https://github.com/NLog/NLog/wiki/Dependency-injection-with-NLog#resolve-dependency-injection-interface-during-object-creation)-delegate.

This should be the stepping stone for moving from NLog v5 to NLog v6. Where NLog v6 will only support the new style. See also #3366

**Comment by snakefoot** [comment]
Step 2 would probably be to focus on not using reflection for property-assignment, where the NLog Configuration Item Type announces what options are available. Ex. implements interface with method that returns dictionary of functors for assigning available properties.

**Comment by snakefoot** [comment]
This should be the stepping stone for moving from NLog v5 to NLog v6. Where NLog v6 will only support the new style.

**Comment by snakefoot** [comment]
Looks like Step 2 can be "easily" reached by just using `typeof(T).GetProperties()` like this:
```c#
public static T RegisterType<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicProperties)] T>(Func<T> targetCreator) where T : TargetBase
{
    var properties = () => typeof(T).GetProperties();
}
```
Was cheated by PolySharp, that apparently doesn't work correctly. PolySharp correctly declared `DynamicallyAccessedMembers` and `DynamicallyAccessedMemberTypes` for the compiler, but the linker didn't react to them. But having them directly in the NLog-assembly worked correctly.

**Comment by snakefoot** [comment]
Resolved by #5191
